tutorial/markdown/python/streamlit/flight_search_streamlit_couchbase.md

```diff
@@ -0,0 +1,709 @@
+
+```python
+@st.cache_data
+def get_all_hotels(_connection, cities):
+    cities_str = "[" # Initialize the string for city names
+    for i in range(len(cities)):
+        if i != len(cities) - 1:
+            cities_str += f'"{cities[i]}", ' # Add city name with a comma
+        else:
+            cities_str += f'"{cities[i]}"' # Add last city without a comma
+    cities_str += "]"
+    query = f"""
+    SELECT h.*, geo.lat as lat, geo.lon as lon, ARRAY_AVG(ARRAY r.ratings.Overall FOR r IN h.reviews WHEN r.ratings.Overall IS NOT MISSING END) as avg_rating
+    FROM `travel-sample`.inventory.hotel h
+    WHERE h.geo.lat IS NOT MISSING 
+    AND h.type = "hotel" 
+    AND h.geo.lon IS NOT MISSING 
+    AND h.city IN {cities_str}
+    """
+    result = _connection.query(query)
+    hotels = []
+    for row in result:
+        hotels.append(row)
+    return pd.DataFrame(hotels)
+```
+
+### Create Interactive Visualizations
+
+`plot_airports_and_routes(airports_df, routes_df)`: Displays airports and their flight routes.
+
+This function visualizes flight routes between airports using [Plotly](https://plotly.com/python/lines-on-tile-maps/). It first extracts airport coordinates from `airports_df` into a dictionary for quick lookup. Then, it iterates through `routes_df` to collect latitude and longitude pairs for each flight route, ensuring that non-existent airports are skipped. A scatter map plot is created using `Scattermap` to represent routes as blue lines. Additionally, a separate scatter plot of airports is overlaid, with markers color-coded in red and displaying airport details on hover. The final visualization is displayed using `st.plotly_chart`.
+
+```python
+def plot_airports_and_routes(airports_df, routes_df):
+    fig = go.Figure()
+    # Create a dictionary mapping FAA codes to latitude and longitude for quick lookup
+    airport_coords = {
+        row["faa"]: (row["lat"], row["lon"])
+        for _, row in airports_df.iterrows()
+        if row["faa"] is not None  # Ensure faa is not null
+    }
+    lats = []
+    lons = []
+    # Iterate through routes to fetch airport coordinates and construct flight paths
+    for _, row in routes_df.iterrows():
+        source_coords = airport_coords.get(row["sourceairport"])
+        dest_coords = airport_coords.get(row["destinationairport"])
+        if source_coords and dest_coords:
+            lats.extend([source_coords[0], dest_coords[0], None])  # None for breaks
+            lons.extend([source_coords[1], dest_coords[1], None])
```

Comment:
This looks like it could be done much more efficiently using Pandas functions. 
There are issues with iterating over the rows in Pandas which you are performing twice here. If at all you need to perform this, please do it once. 
https://stackoverflow.com/questions/16476924/how-can-i-iterate-over-rows-in-a-pandas-dataframe